### PR TITLE
[36136] Rework members page to show name and avatars for principals

### PR DESCRIPTION
**Pull request by oliverguenther**
- [x] Output avatar for all types of principals
- [x] Remove first and lastname, add common name column that remains sortable
- [x] Extend members specs for this change 

https://community.openproject.com/wp/36136

### CODE REVIEW

**Comment by ulferts** [comment]
@oliverguenther #8961 got merged.

**Comment by oliverguenther** [comment]
@ulferts I could use your help in finding out how to properly add the query order option for `name` to correctly use the full name now. There is a filter that does the column handling depending on the name formatting setting, but there's no order yet. That order would need a complex statement or an additional select to an alias that we can then sort on.

I can't seem to find a reference for how to do that other than the work packages `:sortable` definitions which IIRC was not desired either?

File: app/cells/members/row_cell.rb [hunk 1]
```diff
@@ -18,12 +19,17 @@ def row_css_class
...
-      link_to principal.lastname, user_path(principal) if user?
-    end
+    def name
+      icon = avatar principal, class: 'avatar-mini'
+
+      link =
+        if user?
+          link_to principal.name, user_path(principal)
+        else
+          content_tag :span, principal.name
```
Comment (ulferts) [inline app/cells/members/row_cell.rb]: One could link this to the new group show page in case of a group.

File: app/cells/members/table_cell.rb [hunk 2]
```diff
@@ -1,11 +1,11 @@
 module Members
   class TableCell < ::TableCell
-    options :authorize_update, :available_roles
-    columns :lastname, :firstname, :mail, :roles, :groups, :status
-    sortable_columns :lastname, :firstname, :mail
+    options :authorize_update, :available_roles, :is_filtered
+    columns :name, :mail, :roles, :groups, :status
+    sortable_columns :name, :mail
```
Comment (ulferts) [inline app/cells/members/table_cell.rb]: Probably not an issue as I have never heard it to be one before but it is a little bit strange that one cannot sort by status.

File: app/models/queries/members/orders/name_order.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+
+class Queries::Members::Orders::NameOrder < Queries::BaseOrder
+  self.model = Principal
```
Comment (ulferts) [inline app/models/queries/members/orders/name_order.rb]: Normally, this should be `Member` (and actually simply be deduced from the namespace). Then the `left_outer_joins` method would be overwritten to include the principal association.

**Review by ulferts — CHANGES_REQUESTED** [review]
* The sorting by name does not work as empty values for the firstname value for groups and placeholder users lead to those principals being displayed out of order. This did exist before as well but it now becomes even more apparent.:

image


* Just a minor thing but the "Groups" column is currently displaying the placeholder's name for a placeholder user. 

* Existing before but a good opportunity to remove it, the filter section cannot be closed: 

image

Other than that, only small code remarks.

**Comment by oliverguenther** [comment]
@ulferts I believe I addressed all your remarks.

File: app/cells/members/row_cell.rb [hunk 4]
```diff
@@ -149,6 +144,21 @@ def column_css_class(column)
...
     end
 
+    def principal_link
+      case Principal
+      when User
+        link_to principal.name, user_path(principal)
+      when Group
+        link_to principal.name, show_group_path(principal)
+      else
+        content_tag :span, principal.name
```
Comment (ulferts) [inline app/cells/members/row_cell.rb:154]: We need to revisit this once the placeholder users administration PR (#8944) is in.

File: app/models/queries/users/orders/name_order.rb [hunk 5]
```diff
@@ -28,22 +28,8 @@
...
-    if direction == :desc
-      ordered = ordered.reverse_order
-    end
-
-    ordered
-  end
+class Queries::Users::Orders::NameOrder < Queries::Principals::Orders::NameOrder
+  # .user is important here as it forces
+  # "AND users.type = 'User'" which otherwise gets lost for some reason
+  self.model = User.user
```
Comment (ulferts) [inline app/models/queries/users/orders/name_order.rb:34]: This has also bugged me for reasons beyond me and is kinda unsettling.

File: app/models/queries/principals/orders/name_order.rb [hunk 6]
```diff
@@ -35,15 +35,22 @@ def self.key
...
-  def order
-    ordered = self.model.order_by_name
-
-    if direction == :desc
-      ordered = ordered.reverse_order
+      Arel.sql("NULLIF(#{attr}, '') #{direction} NULLS LAST")
     end
+  end
 
-    ordered
```
Comment (ulferts) [inline app/models/queries/principals/orders/name_order.rb:47]: This leads to the groups and placeholder users always being sorted to the end. 

An alternative would be to sort by the combined attributes:

```
Arel.sql("#{User::USER_FORMATS_STRUCTURE[Setting.user_format].join(" || ' ' || ")} #{direction = :desc ? "DESC" : "ASC"}")
```

Don't know about the performance impact, though. Maybe we need a couple of indices in the long run but I'd be willing to try it out without them for now.

File: app/models/principal.rb [hunk 7]
```diff
@@ -86,8 +86,9 @@ def self.search_scope_without_project(project, query)
     not_locked.like(query).not_in_project(project)
   end
 
-  def self.order_by_name
-    order(User::USER_FORMATS_STRUCTURE[Setting.user_format].map { |format| "#{Principal.table_name}.#{format}" })
+  def self.order_by_name(desc: false)
+    OpenProject::Deprecation.warn "Use Queries::Principals with .order('name') instead"
+    order Queries::Principals::Orders::NameOrder.name_order_statements(desc)
```
Comment (ulferts) [inline app/models/principal.rb]: I am not yet decided on all of this and am very open for discussion. Currently I'd rather place more of the functionality in scopes again and use them in the query order objects. 

Scopes seem to be more easy to use throughout the code. This of course is highly subjective.

**Review by ulferts — CHANGES_REQUESTED** [review]
Sorry to be nagging.

File: app/models/queries/principals/orders/name_order.rb (hunk 6 above)
Comment (oliverguenther) [reply app/models/queries/principals/orders/name_order.rb:47]: I did something similar to the user name filter now and used `concat_ws`. I'm quite happy with the solution now

File: app/models/principal.rb (hunk 7 above)
Comment (oliverguenther) [reply app/models/principal.rb]: I replaced it with a separate scope

**Comment by oliverguenther** [comment]
I believe I addressed all remarks that we can right now (Placeholder PR is not yet in)

File: app/models/queries/principals/orders/name_order.rb (hunk 6 above)
Comment (ulferts) [reply app/models/queries/principals/orders/name_order.rb:47]: Works splendidly, thanks.

File: app/models/principal.rb (hunk 7 above)
Comment (ulferts) [reply app/models/principal.rb]: Merci

**Review by ulferts — APPROVED** [review]

**Pull request merged**